crates/sui-core/src/transaction_driver/effects_certifier.rs

```diff
@@ -269,6 +303,100 @@ impl EffectsCertifier {
         }
     }
 
+    /// Gets full effects from a validator, with speculative fallback to other validators.
+    ///
+    /// If the initial validator doesn't respond within GET_FULL_EFFECTS_FALLBACK_DELAY,
+    /// we start parallel requests to validators that have already acknowledged the effects
+    /// (received via the acked_validators channel from wait_for_acknowledgments).
+    ///
+    /// This prevents slow validators from blocking the entire operation when faster
+    /// validators are available, while still preferring the initial validator if it responds quickly.
+    #[instrument(level = "debug", skip_all, fields(tx_digest = ?tx_digest, initial_validator = ?initial_target))]
+    async fn get_full_effects_with_fallback<A>(
+        &self,
+        authority_aggregator: &Arc<AuthorityAggregator<A>>,
+        initial_client: Arc<SafeClient<A>>,
+        initial_target: AuthorityName,
+        tx_digest: Option<TransactionDigest>,
+        tx_type: TxType,
+        consensus_position: Option<ConsensusPosition>,
+        options: &SubmitTransactionOptions,
+        mut acked_validators_rx: Receiver<AuthorityName>,
+    ) -> (FullEffectsResult, AuthorityName)
+    where
+        A: AuthorityAPI + Send + Sync + 'static + Clone,
+    {
+        let mut pending_requests: FuturesUnordered<
+            BoxFuture<'_, (AuthorityName, FullEffectsResult)>,
+        > = FuturesUnordered::new();
+
+        // Add initial request to the pending set alongside fallbacks for uniform handling
+        let initial_request = self.get_full_effects(
+            initial_client,
+            tx_digest,
+            tx_type,
+            consensus_position,
+            options,
+        );
+        pending_requests.push(Box::pin(
+            async move { (initial_target, initial_request.await) },
+        ));
+
+        let mut fallback_delay = tokio::time::interval(GET_FULL_EFFECTS_FALLBACK_DELAY);
+
+        loop {
+            tokio::select! {
+                Some((validator, result)) = pending_requests.next() => {
+                    if validator == initial_target {
+                        // Initial request completed - always return its result (success or error)
+                        return (result, validator);
+                    }
+                    // Fallback request completed
+                    if result.is_ok() {
+                        return (result, validator);
+                    }
```

Comment:
When the txn is rejected, fallbacks fire until the rejection gets certified. A more optimal approach seems to be to not fire fallback request the effect digest is certified. Then ignoring errors make more sense.
Maybe merging the fallback logic into get_certified_finalized_effects() is more ideal. But that function is already a bit complex, so refactoring is needed. We can merge this for now and do a followup.

Reply:
I would agree with that. Probably this method should not try to get a non-error response back. But return as soon as it gets _a_ response back